File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/DocumentRouteSelectorPolicy.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.config.subscription.ConfigSubscriber;
import com.yahoo.document.Document;
import com.yahoo.document.DocumentPut;
import com.yahoo.document.DocumentType;
import com.yahoo.document.select.DocumentSelector;
import com.yahoo.document.select.Result;
import com.yahoo.log.LogLevel;
import com.yahoo.messagebus.Message;
import com.yahoo.messagebus.metrics.MetricSet;
import com.yahoo.messagebus.routing.Route;
import com.yahoo.messagebus.routing.RoutingContext;

import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;

/**
 * This policy is responsible for selecting among the given recipient routes according to the configured document
 * selection properties. To facilitate this the "routing" plugin in the vespa model builds a mapping from the route
 * names to a document selector and a feed name of every search cluster. This can very well be extended to include
 * storage at a later time.
 *
 * @author <a href="mailto:simon@yahoo-inc.com">Simon Thoresen</a>
 */
public class DocumentRouteSelectorPolicy
        implements DocumentProtocolRoutingPolicy, ConfigSubscriber.SingleSubscriber<DocumentrouteselectorpolicyConfig> {

    private static Logger log = Logger.getLogger(DocumentRouteSelectorPolicy.class.getName());
    private Map<String, DocumentSelector> config;
    private String error = "Not configured.";
    private ConfigSubscriber subscriber;

    /**
     * This policy is constructed with a configuration identifier that can be subscribed to for the document selector
     * config. If the string is either null or empty it will default to the proper one.
     *
     * @param configId The configuration identifier to subscribe to.
     */
    public DocumentRouteSelectorPolicy(String configId) {
        subscriber = new ConfigSubscriber();
        subscriber.subscribe(this, DocumentrouteselectorpolicyConfig.class, configId);
    }

    /**
     * This is a safety mechanism to allow the constructor to fail and signal that it can not be used.
     *
     * @return The error string, or null if no error.
     */
    public synchronized String getError() {
        return error;
    }

    /**
     * This method is called when configuration arrives from the config server. The received config object is traversed
     * and a local map is constructed and swapped with the current {@link #config} map.
     *
     * @param cfg The configuration object given by subscription.
     */
    @Override
    public void configure(DocumentrouteselectorpolicyConfig cfg) {
        String error = null;
        Map<String, DocumentSelector> config = new HashMap<>();
        for (int i = 0; i < cfg.route().size(); i++) {
            DocumentrouteselectorpolicyConfig.Route route = cfg.route(i);
            if (route.selector().isEmpty()) {
                continue;
            }
            DocumentSelector selector;
            try {
                selector = new DocumentSelector(route.selector());
                log.log(LogLevel.CONFIG, "Selector for route '" + route.name() + "' is '" + selector + "'.");
            } catch (com.yahoo.document.select.parser.ParseException e) {
                error = "Error parsing selector '" + route.selector() + "' for route '" + route.name() + "; " +
                        e.getMessage();
                break;
            }
            config.put(route.name(), selector);
        }
        synchronized (this) {
            this.config = config;
            this.error = error;
        }
    }

    @Override
    public void select(RoutingContext context) {
        // Require that recipients have been configured.
        if (context.getNumRecipients() == 0) {
            context.setError(DocumentProtocol.ERROR_POLICY_FAILURE,
                             "No recipients configured.");
            return;
        }

        // Invoke private select method for each candidate recipient.
        synchronized (this) {
            if (error != null) {
                context.setError(DocumentProtocol.ERROR_POLICY_FAILURE, error);
                return;
            }
            for (int i = 0; i < context.getNumRecipients(); ++i) {
                Route recipient = context.getRecipient(i);
                String routeName = recipient.toString();
                if (select(context, routeName)) {
                    Route route = context.getMessageBus().getRoutingTable(DocumentProtocol.NAME).getRoute(routeName);
                    context.addChild(route != null ? route : recipient);
                }
            }
        }
        context.setSelectOnRetry(false);

        // Notify that no children were selected, this is to differentiate this from the NO_RECIPIENTS_FOR_ROUTE error
        // that message bus will generate if there are no recipients and no reply.
        if (context.getNumChildren() == 0) {
            context.setReply(new DocumentIgnoredReply());
        }
    }

    /**
     * This method runs the selector associated with the given location on the content of the message. If the selector
     * validates the location, this method returns true.
     *
     * @param context   The routing context that contains the necessary data.
     * @param routeName The candidate route whose selector to run.
     * @return Whether or not to send to the given recipient.
     */
    private boolean select(RoutingContext context, String routeName) {
        if (config == null) {
            return true;
        }
        DocumentSelector selector = config.get(routeName);
        if (selector == null) {
            return true;
        }

        // Select based on message content.
        Message msg = context.getMessage();
        switch (msg.getType()) {

        case DocumentProtocol.MESSAGE_PUTDOCUMENT:
            return selector.accepts(((PutDocumentMessage)msg).getDocumentPut()) == Result.TRUE;

        case DocumentProtocol.MESSAGE_UPDATEDOCUMENT:
            return selector.accepts(((UpdateDocumentMessage)msg).getDocumentUpdate()) != Result.FALSE;

        case DocumentProtocol.MESSAGE_REMOVEDOCUMENT: {
            RemoveDocumentMessage removeMsg = (RemoveDocumentMessage)msg;
            if (removeMsg.getDocumentId().hasDocType()) {
                return selector.accepts(removeMsg.getDocumentRemove()) != Result.FALSE;
            } else {
                return true;
            }
        }

        case DocumentProtocol.MESSAGE_BATCHDOCUMENTUPDATE:
            BatchDocumentUpdateMessage bdu = (BatchDocumentUpdateMessage)msg;
            for (int i = 0; i < bdu.getUpdates().size(); i++) {
                if (selector.accepts(bdu.getUpdates().get(i)) == Result.FALSE) {
                    return false;
                }
            }
            return true;

        default:
            return true;
        }
    }

    @Override
    public void merge(RoutingContext context) {
        DocumentProtocol.merge(context);
    }

    @Override
    public void destroy() {
        if (subscriber != null) {
            subscriber.close();
        }
    }

    @Override
    public MetricSet getMetrics() {
        return null;
    }
}

File: documentapi/src/vespa/documentapi/messagebus/policies/documentrouteselectorpolicy.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "documentrouteselectorpolicy.h"
#include <vespa/document/bucket/bucketidfactory.h>
#include <vespa/document/select/parser.h>
#include <vespa/documentapi/messagebus/documentprotocol.h>
#include <vespa/documentapi/messagebus/messages/getdocumentmessage.h>
#include <vespa/documentapi/messagebus/messages/putdocumentmessage.h>
#include <vespa/documentapi/messagebus/messages/updatedocumentmessage.h>
#include <vespa/documentapi/messagebus/messages/documentignoredreply.h>
#include <vespa/messagebus/emptyreply.h>
#include <vespa/messagebus/routing/routingtable.h>
#include <vespa/messagebus/messagebus.h>
#include <vespa/vespalib/util/stringfmt.h>
#include <vespa/config/helper/configfetcher.hpp>
#include <vespa/config/subscription/configuri.h>
#include <vespa/documentapi/messagebus/messages/removedocumentmessage.h>

#include <vespa/log/log.h>
LOG_SETUP(".documentrouteselectorpolicy");

using document::select::Result;

namespace documentapi {

DocumentRouteSelectorPolicy::DocumentRouteSelectorPolicy(
        const document::DocumentTypeRepo &repo, const config::ConfigUri & configUri) :
    mbus::IRoutingPolicy(),
    config::IFetcherCallback<messagebus::protocol::DocumentrouteselectorpolicyConfig>(),
    _repo(repo),
    _lock(),
    _config(),
    _error("Not configured."),
    _fetcher(std::make_unique<config::ConfigFetcher>(configUri.getContext()))
{
    _fetcher->subscribe<messagebus::protocol::DocumentrouteselectorpolicyConfig>(configUri.getConfigId(), this);
    _fetcher->start();
}

DocumentRouteSelectorPolicy::~DocumentRouteSelectorPolicy() = default;

void
DocumentRouteSelectorPolicy::configure(std::unique_ptr<messagebus::protocol::DocumentrouteselectorpolicyConfig> cfg)
{
    string error = "";
    ConfigMap config;
    for (uint32_t i = 0; i < cfg->route.size(); i++) {
        const messagebus::protocol::DocumentrouteselectorpolicyConfig::Route &route = cfg->route[i];
        if (route.selector.empty()) {
            continue;
        }
        SelectorPtr selector;
        try {
            document::BucketIdFactory factory;
            document::select::Parser parser(_repo, factory);
            selector.reset(parser.parse(route.selector).release());
        }
        catch (document::select::ParsingFailedException &e) {
            error = vespalib::make_string("Error parsing selector '%s' for route '%s'; %s",
                                          route.selector.c_str(), route.name.c_str(),
                                          e.getMessage().c_str());
            break;
        }
        config[string(route.name)] = selector;
    }
    std::lock_guard guard(_lock);
    _config.swap(config);
    _error.swap(error);
}

const string &
DocumentRouteSelectorPolicy::getError() const
{
    std::lock_guard guard(_lock);
    return _error;
}

void
DocumentRouteSelectorPolicy::select(mbus::RoutingContext &context)
{
    // Require that recipients have been configured.
    if (!context.hasRecipients()) {
        context.setError(DocumentProtocol::ERROR_POLICY_FAILURE,
                         "No recipients configured.");
        return;
    }

    // Invoke private select method for each candidate recipient.
    {
        std::lock_guard guard(_lock);
        if (!_error.empty()) {
            context.setError(DocumentProtocol::ERROR_POLICY_FAILURE, _error);
            return;
        }
        for (uint32_t i = 0; i < context.getNumRecipients(); ++i) {
            const mbus::Route &recipient = context.getRecipient(i);
            vespalib::string routeName = recipient.toString();
            if (select(context, routeName)) {
                const mbus::Route *route = context.getMessageBus().getRoutingTable(DocumentProtocol::NAME)->getRoute(routeName);
                context.addChild(route != NULL ? *route : recipient);
            }
        }
    }
    context.setSelectOnRetry(false);

    // Notify that no children were selected, this is to differentiate this from the NO_RECIPIENTS_FOR_ROUTE error
    // that message bus will generate if there are no recipients and no reply.
    if (!context.hasChildren()) {
        context.setReply(mbus::Reply::UP(new DocumentIgnoredReply()));
    }
}

bool
DocumentRouteSelectorPolicy::select(mbus::RoutingContext &context, const vespalib::string &routeName)
{
    if (_config.empty()) {
        LOG(debug, "No config at all, select '%s'.", routeName.c_str());
        return true;
    }
    ConfigMap::const_iterator it = _config.find(routeName);
    if (it == _config.end()) {
        LOG(debug, "No config entry for route '%s', select it.", routeName.c_str());
        return true;
    }
    LOG_ASSERT(it->second.get() != NULL);

    // Select based on message content.
    const mbus::Message &msg = context.getMessage();
    switch(msg.getType()) {
    case DocumentProtocol::MESSAGE_PUTDOCUMENT:
        return it->second->contains(static_cast<const PutDocumentMessage&>(msg).getDocument()) == Result::True;

    case DocumentProtocol::MESSAGE_UPDATEDOCUMENT:
        return it->second->contains(static_cast<const UpdateDocumentMessage&>(msg).getDocumentUpdate()) != Result::False;

    case DocumentProtocol::MESSAGE_REMOVEDOCUMENT: {
        const RemoveDocumentMessage &removeMsg = static_cast<const RemoveDocumentMessage &>(msg);
        if (removeMsg.getDocumentId().hasDocType()) {
            return it->second->contains(removeMsg.getDocumentId()) != Result::False;
        } else {
            return true;
        }
    }
    case DocumentProtocol::MESSAGE_GETDOCUMENT: {
        const GetDocumentMessage &getMsg = static_cast<const GetDocumentMessage &>(msg);
        if (getMsg.getDocumentId().hasDocType()) {
            return it->second->contains(getMsg.getDocumentId()) != Result::False;
        } else {
            return true;
        }
    }
    default:
        return true;
    }
}

void
DocumentRouteSelectorPolicy::merge(mbus::RoutingContext &context)
{
    DocumentProtocol::merge(context);
}

}
